File: ratelimit/redis/backend.go
// Package redislimit implements a shared rolling-window limiter using Redis Lua.
package redislimit

import (
	"context"
	"crypto/sha256"
	_ "embed"
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/proven-trade/cex-sdk/v2/ratelimit"
	"github.com/redis/go-redis/v9"
)

//go:embed backend.lua
var scriptSource string

var script = redis.NewScript(scriptSource)

const (
	maxUnits    = 1<<31 - 1
	maxDuration = time.Duration(1<<63-1) / time.Microsecond * time.Microsecond
)

// Config defines the Redis connection and the namespace shared by SDK processes.
type Config struct {
	// Client is owned by the caller. ContextTimeoutEnabled must be true.
	Client redis.UniversalClient
	// Namespace must match across processes sharing a quota. Use stable account
	// and route IDs in SDK configurations as well.
	Namespace string
	// OperationTimeout bounds each Redis call, including methods without a context.
	// It defaults to two seconds.
	OperationTimeout time.Duration
}

// Backend keeps all keys for a namespace in a single Redis Cluster hash slot.
// Redis server time is authoritative. Storage errors never grant quota.
type Backend struct {
	client  redis.UniversalClient
	prefix  string
	timeout time.Duration
}

var _ ratelimit.Backend = (*Backend)(nil)

// New creates a backend without connecting to Redis. The caller owns Client.Close.
func New(config Config) (*Backend, error) {
	if config.Client == nil {
		return nil, fmt.Errorf("Redis client is required")
	}
	var respectsContext bool
	switch client := config.Client.(type) {
	case *redis.Client:
		respectsContext = client != nil && client.Options().ContextTimeoutEnabled
	case *redis.ClusterClient:
		respectsContext = client != nil && client.Options().ContextTimeoutEnabled
	case *redis.Ring:
		respectsContext = client != nil && client.Options().ContextTimeoutEnabled
	default:
		return nil, fmt.Errorf("Redis limiter requires a standard Redis client, cluster, or ring")
	}
	if !respectsContext {
		return nil, fmt.Errorf("Redis limiter requires ContextTimeoutEnabled")
	}
	config.Namespace = strings.TrimSpace(config.Namespace)
	if config.Namespace == "" {
		return nil, fmt.Errorf("Redis limiter namespace is required")
	}
	if config.OperationTimeout < 0 {
		return nil, fmt.Errorf("Redis operation timeout cannot be negative")
	}
	if config.OperationTimeout == 0 {
		config.OperationTimeout = 2 * time.Second
	}
	return &Backend{client: config.Client, prefix: fmt.Sprintf("cex-sdk:ratelimit:{%x}:", sha256.Sum256([]byte(config.Namespace))), timeout: config.OperationTimeout}, nil
}

func (backend *Backend) keys(names []string) []string {
	keys := make([]string, 0, len(names)*2)
	for _, name := range names {
		base := backend.prefix + fmt.Sprintf("%x", sha256.Sum256([]byte(name)))
		keys = append(keys, base+":rule", base+":events")
	}
	return keys
}

func (backend *Backend) call(ctx context.Context, names []string, args ...any) ([]int64, error) {
	ctx, cancel := context.WithTimeout(ctx, backend.timeout)
	defer cancel()
	values, err := script.Run(ctx, backend.client, backend.keys(names), args...).Int64Slice()
	if err != nil {
		// Preserve context identity without rendering connection credentials or URLs.
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		return nil, fmt.Errorf("Redis limiter operation failed: %w", err)
	}
	if len(values) == 0 {
		return nil, fmt.Errorf("Redis limiter returned an empty response")
	}
	switch values[0] {
	case -1:
		return nil, ratelimit.ErrUnknownRule
	case -2:
		return nil, ratelimit.ErrInvalidRule
	}
	return values, nil
}

// SetRule preserves usage and cooldown when a rule is updated. A window change
// blocks for a full new window because previously expired history is unavailable.
func (backend *Backend) SetRule(rule ratelimit.Rule) error {
	return backend.SetRuleContext(context.Background(), rule)
}

// SetRuleContext registers a rule within the caller's deadline. OperationTimeout
// is an additional upper bound; it never extends the request budget.
func (backend *Backend) SetRuleContext(ctx context.Context, rule ratelimit.Rule) error {
	if ctx == nil {
		return fmt.Errorf("limiter context cannot be nil")
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	rule.Key = strings.TrimSpace(rule.Key)
	if rule.Key == "" || rule.Limit <= 0 || rule.Limit > maxUnits || rule.Window <= 0 || rule.Window > maxDuration {
		return ratelimit.ErrInvalidRule
	}
	_, err := backend.call(ctx, []string{rule.Key}, "set", rule.Limit, durationMicros(rule.Window))
	return err
}

// Wait checks and charges every dimension atomically, waiting with cancellation.
func (backend *Backend) Wait(ctx context.Context, charges ...ratelimit.Charge) error {
	if ctx == nil {
		return fmt.Errorf("limiter context cannot be nil")
	}
	combined := make(map[string]int)
	for _, charge := range charges {
		key := strings.TrimSpace(charge.Key)
		if key == "" || charge.Units <= 0 || charge.Units > maxUnits-combined[key] {
			return ratelimit.ErrInvalidRule
		}
		combined[key] += charge.Units
	}
	names := make([]string, 0, len(combined))
	for name := range combined {
		names = append(names, name)
	}
	sort.Strings(names)
	args := []any{"wait"}
	for _, name := range names {
		args = append(args, combined[name])
	}
	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if len(names) == 0 {
			return nil
		}
		values, err := backend.call(ctx, names, args...)
		if err != nil {
			return err
		}
		if values[0] == 0 {
			return nil
		}
		timer := time.NewTimer(time.Duration(values[0]) * time.Microsecond)
		select {
		case <-ctx.Done():
			timer.Stop()
			return ctx.Err()
		case <-timer.C:
		}
	}
}

// ObserveUsed conservatively raises shared usage to the server-observed value.
func (backend *Backend) ObserveUsed(key string, used int) error {
	if used < 0 || used > maxUnits {
		return ratelimit.ErrInvalidRule
	}
	_, err := backend.call(context.Background(), []string{key}, "observe", used)
	return err
}

// BlockFor extends shared cooldowns without shortening an existing block.
func (backend *Backend) BlockFor(keys []string, duration time.Duration) error {
	if duration <= 0 || duration > maxDuration {
		return ratelimit.ErrInvalidRule
	}
	if len(keys) == 0 {
		return nil
	}
	_, err := backend.call(context.Background(), keys, "block", durationMicros(duration))
	return err
}

// Snapshot returns state using the same server clock as acquisition.
func (backend *Backend) Snapshot(key string) (ratelimit.Snapshot, error) {
	values, err := backend.call(context.Background(), []string{key}, "snapshot")
	if err != nil {
		return ratelimit.Snapshot{}, err
	}
	if len(values) != 6 {
		return ratelimit.Snapshot{}, errors.New("invalid Redis limiter snapshot")
	}
	result := ratelimit.Snapshot{Rule: ratelimit.Rule{Key: key, Limit: int(values[1]), Window: time.Duration(values[2]) * time.Microsecond}, Used: int(values[3])}
	if values[4] > 0 {
		result.WindowStart = time.UnixMicro(values[4])
	}
	if values[5] > 0 {
		result.BlockedUntil = time.UnixMicro(values[5])
	}
	return result, nil
}

func durationMicros(value time.Duration) int64 {
	micros := value.Microseconds()
	if value%time.Microsecond != 0 {
		micros++
	}
	return micros
}

File: ratelimit/redis/backend.lua
-- Every rule/events key is supplied in KEYS and shares a namespace hash slot.
-- Responses: {-1} unknown rule, {-2} invalid charge, {wait microseconds}.
local clock = redis.call('TIME')
local now = tonumber(clock[1]) * 1000000 + tonumber(clock[2])
local action = ARGV[1]

if action == 'set' then
    local oldWindow = tonumber(redis.call('HGET', KEYS[1], 'window'))
    local window = tonumber(ARGV[3])
    if oldWindow and oldWindow ~= window then
        local blocked = tonumber(redis.call('HGET', KEYS[1], 'blocked') or '0')
        redis.call('HSET', KEYS[1], 'blocked', math.max(blocked, now + window))
    end
    redis.call('HSET', KEYS[1], 'limit', ARGV[2], 'window', window)
    return {0}
end

local states = {}
-- Validate the entire request before mutating any quota dimension.
for i = 1, #KEYS, 2 do
    local fields = redis.call('HMGET', KEYS[i], 'limit', 'window', 'used', 'blocked')
    if not fields[1] then return {-1} end
    local state = {
        key = KEYS[i], events = KEYS[i+1], limit = tonumber(fields[1]),
        window = tonumber(fields[2]), used = tonumber(fields[3] or '0'),
        blocked = tonumber(fields[4] or '0')
    }
    if action == 'wait' then
        state.charge = tonumber(ARGV[(i+1)/2 + 1])
        if not state.charge or state.charge <= 0 or state.charge > state.limit then return {-2} end
    end
    states[#states+1] = state
end

local function units(member)
    return tonumber(string.match(member, ':(%d+)$'))
end

local function prune(state)
    local expired = redis.call('ZRANGEBYSCORE', state.events, '-inf', now - state.window)
    for _, member in ipairs(expired) do state.used = state.used - units(member) end
    if #expired > 0 then
        redis.call('ZREMRANGEBYSCORE', state.events, '-inf', now - state.window)
        redis.call('HSET', state.key, 'used', state.used)
    end
end

local function charge(state, amount)
    local sequence = redis.call('HINCRBY', state.key, 'sequence', 1)
    redis.call('ZADD', state.events, now, sequence .. ':' .. amount)
    state.used = state.used + amount
    redis.call('HSET', state.key, 'used', state.used)
end

for _, state in ipairs(states) do prune(state) end

if action == 'wait' then
    local waitUntil = now
    for _, state in ipairs(states) do
        waitUntil = math.max(waitUntil, state.blocked)
        local excess = state.used + state.charge - state.limit
        if excess > 0 then
            local events = redis.call('ZRANGE', state.events, 0, -1, 'WITHSCORES')
            for i = 1, #events, 2 do
                excess = excess - units(events[i])
                if excess <= 0 then
                    waitUntil = math.max(waitUntil, tonumber(events[i+1]) + state.window)
                    break
                end
            end
        end
    end
    if waitUntil > now then return {waitUntil - now} end
    for _, state in ipairs(states) do charge(state, state.charge) end
    return {0}
elseif action == 'observe' then
    local state = states[1]
    local used = tonumber(ARGV[2])
    if used > state.used then charge(state, used - state.used) end
    return {0}
elseif action == 'block' then
    for _, state in ipairs(states) do
        redis.call('HSET', state.key, 'blocked', math.max(state.blocked, now + tonumber(ARGV[2])))
    end
    return {0}
elseif action == 'snapshot' then
    local state = states[1]
    local first = redis.call('ZRANGE', state.events, 0, 0, 'WITHSCORES')
    return {0, state.limit, state.window, state.used, tonumber(first[2] or '0'), state.blocked}
end
return redis.error_reply('unsupported limiter operation')
